kernel_tuner/observers/tegra.py

```diff
@@ -0,0 +1,208 @@
+
+    @staticmethod
+    def get_dev_path():
+        """Get the path to device core clock control in /sys"""
+        # loop to find GPU device name based on jetson_clocks
+        for dev in Path("/sys/class/devfreq").iterdir():
+            with open(dev / Path("device/of_node/name")) as fp:
+                name = fp.read().strip().rstrip("\x00")
+            if name in ("gv11b", "gp10b", "ga10b", "gpu"):
+                root_path = dev
+                break
+        else:
+            raise FileNotFoundError("No internal tegra GPU found")
+        return root_path
+
+    def _read_railgate_file(self):
+        """Read railgate status"""
+        with open(self.dev_path / Path("device/railgate_enable")) as fp:
+            data = int(fp.read().strip())
+        return data
+
+    def _write_railgate_file(self, value):
+        """Set railgate status"""
+        if value not in (0, 1):
+            raise ValueError(f"Illegal governor value {value}, must be 0 or 1")
+        full_path = self.dev_path / Path("device/railgate_enable")
+        args = [
+            "sudo",
+            "sh",
+            "-c",
+            f"echo {value} > {str(full_path)}"
+        ]
+        subprocess.run(args, check=True)
+
+    def _read_clock_file(self, fname):
+        """Read current or available frequency value(s) from a frequency control file"""
+        with open(self.dev_path / Path(fname)) as fp:
+            raw_data = np.array(fp.read().strip().split())
+        if len(raw_data) > 1:
+            data = raw_data.astype(int)
+        else:
+            data = int(raw_data)
+        return data
+
+    def _write_clock_file(self, fname, value):
+        """Write a frequency value to a core clock control file"""
+        self.has_changed_clocks = True
+        available_files = ("min_freq", "max_freq")
+        if fname not in available_files:
+            raise ValueError(f"Illegal filename value: {fname}, must be one of {available_files}")
+
+        if value not in self.supported_gr_clocks:
+            raise ValueError(f"Illegal frequency value {value}, must be one of {self.supported_gr_clocks}")
+
+        full_path = self.dev_path / Path(fname)
+        args = [
+            "sudo",
+            "sh",
+            "-c",
+            f"echo {value} > {str(full_path)}"
+        ]
+        subprocess.run(args, check=True)
+
+    @property
+    def gr_clock(self):
+        """Control the core clock frequency"""
+        return self._read_clock_file("cur_freq")
+
+    @gr_clock.setter
+    def gr_clock(self, new_clock):
+        self._write_railgate_file(0)
+        cur_clock = self._read_clock_file("cur_freq")
+        if new_clock > cur_clock:
+            self._write_clock_file("max_freq", new_clock)
+            self._write_clock_file("min_freq", new_clock)
+        elif new_clock < cur_clock:
+            self._write_clock_file("min_freq", new_clock)
+            self._write_clock_file("max_freq", new_clock)
```

Comment:
@loostrum I didn't really understand this bit of code. It seems to me both the if and the else do the same? We could just check if cur_clock != new_clock, or am I missing something?

Reply:
It's about the order of writing min/max. Writing a max lower than the current min doesn't work, the system just ignores it. Same when writing a min higher than the current max.

Reply:
Ah that explains! I'll add a comment in the code to document this, otherwise someone may try to simplify the code and rediscover this behavior!